antimev: 4 nodes crash with error "panic: failed to retrieve next block validators: failed to perform 'getCurrentConsensus' call: execution aborted (timeout = 5s)"

#### System information

Version: 0.3.0-unstable
Git Commit: 20d7ab409f31bd36bf8e73014f90107ca2c43965
Git Commit Date: 20241203
Architecture: amd64
Go Version: go1.22.2
Operating System: linux
GOPATH=
GOROOT=

#### Expected behaviour
Should work without panic error

#### Actual behaviour
4 nodes crash and stop working with error 

#### Steps to reproduce the behaviour
The issue is related to this in development PR https://github.com/bane-labs/go-ethereum/pull/332
1. start seven privnet with cmd
`make privnet_start_seven`
2. start a new node in CN mode, and then voted it to be the new CN (the step maybe optinal)
3. Keep the privnet work running one night

#### Backtrace
```
node1
...
2024-12-04T01:06:16.566+0800	DEBUG	dbft@v0.3.1-0.20241128121733-94bc38235e44/check.go:111	not enough to commit	{"count": 4}
2024-12-04T01:06:16.566+0800	DEBUG	dbft@v0.3.1-0.20241128121733-94bc38235e44/dbft.go:231	received message	{"type": "Commit", "from": 0, "height": 3044, "view": 0, "my_height": 3044, "my_view": 0}
2024-12-04T01:06:16.738+0800	DEBUG	dbft@v0.3.1-0.20241128121733-94bc38235e44/dbft.go:231	received message	{"type": "Commit", "from": 3, "height": 3044, "view": 0, "my_height": 3044, "my_view": 0}
2024-12-04T01:06:16.778+0800	INFO	dbft@v0.3.1-0.20241128121733-94bc38235e44/dbft.go:588	received Commit	{"validator": 3}
2024-12-04T01:06:16.805+0800	INFO	dbft@v0.3.1-0.20241128121733-94bc38235e44/check.go:120	approving block	{"height": 3044, "hash": "0xd679ef4ba453aa0323d6b101e27799523fda67c6fa51dfe8be58b2e46c7615c8", "tx_count": 0, "merkle": "0xf8c71ec88104109d3af57cfdaac1a33ffacd3a7e5ba9e92fb0d7f9620f921527", "prev": "0xc42c5c2a31a57c42932e62a3873e4d52fa37155a572442ea6d7764ba5fe638cd"}
INFO [12-04|01:06:17.456] Chain reorg detected                     number=3042 hash=0a6d8d..e912cf drop=1 dropfrom=cadf91..7e5a02 add=2 addfrom=385596..c36a9b
INFO [12-04|01:06:21.950] Successfully wrote new block with state  number=3044 hash=385596..c36a9b
INFO [12-04|01:06:22.874] New block in the chain                   "dbft index"=3044 "chain index"=3044 hash=0x38559698d9bd5597c9493f449d04d766d2f56d599aa5ba879844a65d46c36a9b "parent hash"=0xc42c5c2a31a57c42932e62a3873e4d52fa37155a572442ea6d7764ba5fe638cd primary=6 coinbase=0x1212000000000000000000000000000000000003 "mix digest"=0x808e594003e9050fb7225a9d410a3bb3249e67f5a25c8f8e620eab217c600494
ERROR[12-04|01:06:36.701] handleDKG error                          height=3044 err="failed to call epochDuration: failed to call at state 'epochDuration': execution aborted (timeout = 5s)"
INFO [12-04|01:06:38.290] Fetching latest sealing proposal         "desired number"=3045
INFO [12-04|01:07:32.683] Commit new sealing work                  number=3045 sealhash=58ffa8..2993e1 "parent hash"=385596..c36a9b etherbase=0x1212000000000000000000000000000000000003 txs=0  gas=0          fees=0          elapsed=1m9.900s
INFO [12-04|01:07:33.368] Sealing proposal updated                 number=3045 sealhash=58ffa8..2993e1 "parent hash"=385596..c36a9b txs=0
panic: failed to retrieve next block validators: failed to perform 'getCurrentConsensus' call: execution aborted (timeout = 5s)

goroutine 4519 [running]:
github.com/ethereum/go-ethereum/consensus/dbft.New.func4({0x0, 0x66d7049d443f49c9?, 0x87baa59a596df5d2?})
	github.com/ethereum/go-ethereum/consensus/dbft/dbft.go:373 +0x216
github.com/nspcc-dev/dbft.(*Context[...]).reset(0x1fcfd00, 0x0, 0x34a9e590ab2dc901)
	github.com/nspcc-dev/dbft@v0.3.1-0.20241128121733-94bc38235e44/context.go:243 +0xdc
github.com/nspcc-dev/dbft.(*DBFT[...]).initializeConsensus(0x1fd3460, 0x0, 0x4)
	github.com/nspcc-dev/dbft@v0.3.1-0.20241128121733-94bc38235e44/dbft.go:93 +0x45
github.com/nspcc-dev/dbft.(*DBFT[...]).Reset(...)
	github.com/nspcc-dev/dbft@v0.3.1-0.20241128121733-94bc38235e44/dbft.go:89
github.com/ethereum/go-ethereum/consensus/dbft.(*DBFT).handleChainBlock(0xc0003186c8, 0xc0082fd408, 0x38?)
	github.com/ethereum/go-ethereum/consensus/dbft/dbft.go:1977 +0xa96
github.com/ethereum/go-ethereum/consensus/dbft.(*DBFT).eventLoop(0xc0003186c8)
	github.com/ethereum/go-ethereum/consensus/dbft/dbft.go:1790 +0xbc5
created by github.com/ethereum/go-ethereum/consensus/dbft.(*DBFT).Start in goroutine 4535
	github.com/ethereum/go-ethereum/consensus/dbft/dbft.go:1572 +0x476


node2

INFO [12-04|01:06:27.476] Looking for peers                        peercount=8 tried=0 static=0
ERROR[12-04|01:06:30.434] handleDKG error                          height=3044 err="failed to call epochDuration: failed to call at state 'epochDuration': execution aborted (timeout = 5s)"
INFO [12-04|01:06:31.914] Fetching latest sealing proposal         "desired number"=3045
...

2024-12-04T10:12:07.630+0800	DEBUG	dbft@v0.3.1-0.20241128121733-94bc38235e44/dbft.go:231	received message	{"type": "RecoveryRequest", "from": 6, "height": 4933, "view": 0, "my_height": 4933, "my_view": 0}


node3
...
ERROR[12-04|01:06:34.049] handleDKG error                          height=3044 err="failed to call epochDuration: failed to call at state 'epochDuration': execution aborted (timeout = 5s)"
...
2024-12-04T10:12:07.606+0800	DEBUG	dbft@v0.3.1-0.20241128121733-94bc38235e44/send.go:8	broadcasting message	{"type": "RecoveryRequest", "height": 4933, "view": 0}




node4
...
ERROR[12-04|01:06:51.784] handleDKG error                          height=3044 err="failed to call epochDuration: failed to call at state 'epochDuration': execution aborted (timeout = 5s)"
...
INFO [12-04|10:10:25.789] Sealing proposal updated                 number=4933 sealhash=6786e6..4e1ebd "parent hash"=e5ed01..472e4d txs=3
panic: failed to retrieve next block validators: failed to perform 'getCurrentConsensus' call: execution aborted (timeout = 5s)

goroutine 4460 [running]:
github.com/ethereum/go-ethereum/consensus/dbft.New.func4({0x0, 0x162730dba9a51547?, 0xc3c6663f0e3f5581?})
	github.com/ethereum/go-ethereum/consensus/dbft/dbft.go:373 +0x216
github.com/nspcc-dev/dbft.(*Context[...]).reset(0x1fcfd00, 0x0, 0xc23687306e572714)
	github.com/nspcc-dev/dbft@v0.3.1-0.20241128121733-94bc38235e44/context.go:243 +0xdc
github.com/nspcc-dev/dbft.(*DBFT[...]).initializeConsensus(0x1fd3460, 0x0, 0x4d2e47d3187b757a)
	github.com/nspcc-dev/dbft@v0.3.1-0.20241128121733-94bc38235e44/dbft.go:93 +0x45
github.com/nspcc-dev/dbft.(*DBFT[...]).Reset(...)
	github.com/nspcc-dev/dbft@v0.3.1-0.20241128121733-94bc38235e44/dbft.go:89
github.com/ethereum/go-ethereum/consensus/dbft.(*DBFT).handleChainBlock(0xc00057ad88, 0xc007650f08, 0x38?)
	github.com/ethereum/go-ethereum/consensus/dbft/dbft.go:1977 +0xa96
github.com/ethereum/go-ethereum/consensus/dbft.(*DBFT).eventLoop(0xc00057ad88)
	github.com/ethereum/go-ethereum/consensus/dbft/dbft.go:1790 +0xbc5
created by github.com/ethereum/go-ethereum/consensus/dbft.(*DBFT).Start in goroutine 4520
	github.com/ethereum/go-ethereum/consensus/dbft/dbft.go:1572 +0x476


node5
...
ERROR[12-04|01:06:33.675] handleDKG error                          height=3044 err="failed to call epochDuration: failed to call at state 'epochDuration': execution aborted (timeout = 5s)"
...
panic: failed to retrieve next block validators: failed to perform 'getCurrentConsensus' call: execution aborted (timeout = 5s)

goroutine 4549 [running]:
github.com/ethereum/go-ethereum/consensus/dbft.New.func4({0x0, 0x7b59be38c3289c3b?, 0x545ee630e07377e1?})
	github.com/ethereum/go-ethereum/consensus/dbft/dbft.go:373 +0x216
github.com/nspcc-dev/dbft.(*Context[...]).reset(0x1fcfd00, 0x0, 0x2da7d6033ed11e18)
	github.com/nspcc-dev/dbft@v0.3.1-0.20241128121733-94bc38235e44/context.go:243 +0xdc
github.com/nspcc-dev/dbft.(*DBFT[...]).initializeConsensus(0x1fd3460, 0x0, 0xe736857a46d7f27a)
	github.com/nspcc-dev/dbft@v0.3.1-0.20241128121733-94bc38235e44/dbft.go:93 +0x45
github.com/nspcc-dev/dbft.(*DBFT[...]).Reset(...)
	github.com/nspcc-dev/dbft@v0.3.1-0.20241128121733-94bc38235e44/dbft.go:89
github.com/ethereum/go-ethereum/consensus/dbft.(*DBFT).handleChainBlock(0xc00023e6c8, 0xc00e62ef08, 0x38?)
	github.com/ethereum/go-ethereum/consensus/dbft/dbft.go:1977 +0xa96
github.com/ethereum/go-ethereum/consensus/dbft.(*DBFT).eventLoop(0xc00023e6c8)
	github.com/ethereum/go-ethereum/consensus/dbft/dbft.go:1790 +0xbc5
created by github.com/ethereum/go-ethereum/consensus/dbft.(*DBFT).Start in goroutine 4497
	github.com/ethereum/go-ethereum/consensus/dbft/dbft.go:1572 +0x476



node6

...
INFO [12-04|10:10:20.028] Sealing proposal updated                 number=4933 sealhash=f453f2..ac4c06 "parent hash"=e62d07..8536e7 txs=3
panic: failed to retrieve next block validators: failed to perform 'getCurrentConsensus' call: execution aborted (timeout = 5s)

goroutine 4536 [running]:
github.com/ethereum/go-ethereum/consensus/dbft.New.func4({0x0, 0x7b59be38c3289c3b?, 0x545ee630e07377e1?})
	github.com/ethereum/go-ethereum/consensus/dbft/dbft.go:373 +0x216
github.com/nspcc-dev/dbft.(*Context[...]).reset(0x1fcfd00, 0x0, 0x6a27bf58848c18bb)
	github.com/nspcc-dev/dbft@v0.3.1-0.20241128121733-94bc38235e44/context.go:243 +0xdc
github.com/nspcc-dev/dbft.(*DBFT[...]).initializeConsensus(0x1fd3460, 0x0, 0xe736857a46d7f27a)
	github.com/nspcc-dev/dbft@v0.3.1-0.20241128121733-94bc38235e44/dbft.go:93 +0x45
github.com/nspcc-dev/dbft.(*DBFT[...]).Reset(...)
	github.com/nspcc-dev/dbft@v0.3.1-0.20241128121733-94bc38235e44/dbft.go:89
github.com/ethereum/go-ethereum/consensus/dbft.(*DBFT).handleChainBlock(0xc000556b48, 0xc00cbb8788, 0x38?)
	github.com/ethereum/go-ethereum/consensus/dbft/dbft.go:1977 +0xa96
github.com/ethereum/go-ethereum/consensus/dbft.(*DBFT).eventLoop(0xc000556b48)
	github.com/ethereum/go-ethereum/consensus/dbft/dbft.go:1790 +0xbc5
created by github.com/ethereum/go-ethereum/consensus/dbft.(*DBFT).Start in goroutine 4502
	github.com/ethereum/go-ethereum/consensus/dbft/dbft.go:1572 +0x476


node7

...
2024-12-04T10:12:07.647+0800	DEBUG	dbft@v0.3.1-0.20241128121733-94bc38235e44/dbft.go:231	received message	{"type": "RecoveryRequest", "from": 6, "height": 4933, "view": 0, "my_height": 4933, "my_view": 0}



node9

...
ERROR[12-04|01:07:32.454] handleDKG error                          height=3044 err="failed to call sharePeriodDuration: failed to call at state 'sharePeriodDuration': execution aborted (timeout = 5s)"
...
2024-12-04T10:12:07.651+0800	DEBUG	dbft@v0.3.1-0.20241128121733-94bc38235e44/send.go:8	broadcasting message	{"type": "RecoveryMessage", "height": 4933, "view": 0}
```
@txhsl @chenquanyu

--- a/consensus/dbft/dbft.go
+++ b/consensus/dbft/dbft.go
@@ -31,6 +31,7 @@ import (
 	"time"
 
 	bls12381 "github.com/consensys/gnark-crypto/ecc/bls12-381"
+	"github.com/ethereum/go-ethereum/accounts/abi"
 	"github.com/ethereum/go-ethereum/antimev"
 	"github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/common/hexutil"
@@ -45,6 +46,7 @@ import (
 	"github.com/ethereum/go-ethereum/core/txpool"
 	"github.com/ethereum/go-ethereum/core/txpool/legacypool"
 	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/core/vm"
 	"github.com/ethereum/go-ethereum/crypto"
 	"github.com/ethereum/go-ethereum/crypto/tpke"
 	"github.com/ethereum/go-ethereum/eth/downloader"
@@ -204,7 +206,7 @@ type DBFT struct {
 	finished chan struct{}
 
 	// various native contract APIs that dBFT uses.
-	ethAPI          *ethapi.BlockChainAPI
+	backend         *ethapi.Backend
 	txAPI           *ethapi.TransactionAPI
 	validatorsCache *lru.Cache[uint64, []common.Address]
 
@@ -1177,14 +1179,11 @@ func (c *DBFT) getBlockWitness(pub *tpke.PublicKey, block *Block) ([]byte, error
 	}
 }
 
-// WithEthAPI initializes Eth blockchain API for proper consensus module work.
-func (c *DBFT) WithEthAPI(api *ethapi.BlockChainAPI) {
-	c.ethAPI = api
-}
-
-// WithTxAPI initializes Eth transaction API for sending transaction.
-func (c *DBFT) WithTxAPI(api *ethapi.TransactionAPI) {
-	c.txAPI = api
+// WithEthAPIBackend initializes Eth API backend and transaction API for
+// proper consensus module work.
+func (c *DBFT) WithEthAPIBackend(b ethapi.Backend) {
+	c.backend = &b
+	c.txAPI = ethapi.NewTransactionAPI(b, new(ethapi.AddrLocker))
 }
 
 // EnforceECDSASignatures enforces ECDSA multisignature block signing scheme.
@@ -2406,8 +2405,8 @@ func (c *DBFT) getValidatorsSorted(blockNum *uint64, state *state.StateDB, heade
 // order used by Governance contract. This method uses cached values in case of
 // validators requested by block height.
 func (c *DBFT) getValidators(blockNum *uint64, state *state.StateDB, header *types.Header) ([]common.Address, error) {
-	if c.ethAPI == nil {
-		return nil, errors.New("eth blockchain API is not initialized, dBFT can't function properly")
+	if c.backend == nil {
+		return nil, errors.New("eth API backend is not initialized, dBFT can't function properly")
 	}
 
 	if state == nil && blockNum != nil {
@@ -2434,23 +2433,22 @@ func (c *DBFT) getValidators(blockNum *uint64, state *state.StateDB, header *typ
 	ctx, cancel := context.WithCancel(context.Background())
 	// Cancel when we are finished consuming integers.
 	defer cancel()
-	var result hexutil.Bytes
+	var result *core.ExecutionResult
 	if state != nil {
-		result, err = c.ethAPI.CallAtState(ctx, args, state, header)
+		result, err = ethapi.DoCallAtState(ctx, *c.backend, args, state, header, nil, nil, 0, 0)
 	} else if blockNum != nil {
 		blockNr := rpc.BlockNumberOrHashWithNumber(rpc.BlockNumber(*blockNum))
-		result, err = c.ethAPI.Call(ctx, args, &blockNr, nil, nil)
+		result, err = ethapi.DoCall(ctx, *c.backend, args, blockNr, nil, nil, 0, 0)
 	} else {
 		return nil, fmt.Errorf("failed to compute validators: both block number and state are nil")
 	}
 	if err != nil {
 		return nil, fmt.Errorf("failed to perform '%s' call: %w", method, err)
 	}
-
 	var res []common.Address
-	err = systemcontracts.GovernanceABI.UnpackIntoInterface(&res, method, result)
+	err = unpackContractExecutionResult(&res, result, systemcontracts.GovernanceABI, method)
 	if err != nil {
-		return nil, fmt.Errorf("failed to unpack validators: %w", err)
+		return nil, err
 	}
 
 	// Update cache in case if existing state was used for validators retrieval.
@@ -2527,4 +2525,16 @@ func (c *DBFT) getNextConsensus(h *types.Header, s *state.StateDB) (common.Hash,
 
 func (c *DBFT) shouldUpdateCommitteeAt(blockNum uint64) bool {
 	return blockNum%uint64(len(c.config.StandByValidators)) == 0
+}
+
+func unpackContractExecutionResult(res interface{}, result *core.ExecutionResult, contractAbi abi.ABI, method string) error {
+	if len(result.Revert()) > 0 {
+		reason, errUnpack := abi.UnpackRevert(result.Revert())
+		if errUnpack == nil {
+			return fmt.Errorf("%w: %v", vm.ErrExecutionReverted, reason)
+		} else {
+			return fmt.Errorf("%w, failed to unpack revert reason: %w", vm.ErrExecutionReverted, errUnpack)
+		}
+	}
+	return contractAbi.UnpackIntoInterface(&res, method, result.Return())
 }

--- a/consensus/dbft/dkg.go
+++ b/consensus/dbft/dkg.go
@@ -52,13 +52,13 @@ type Snapshot struct {
 func (c *DBFT) newSnapshot(h *types.Header, state *state.StateDB, height uint64) (*Snapshot, error) {
 	snap := &Snapshot{}
 	snap.EpochStartHeight = height
-	round, err := getRoundNumber(c.ethAPI, state, h)
+	round, err := getRoundNumber(c.backend, state, h)
 	if err != nil {
 		return nil, err
 	}
 	// Snapshot round index points to the new round, so plus 1
 	snap.Round = round + 1
-	snap.CurrentCNs, err = getCurrentConsensus(c.ethAPI, state, h)
+	snap.CurrentCNs, err = getCurrentConsensus(c.backend, state, h)
 	if err != nil {
 		return nil, err
 	}
@@ -86,11 +86,11 @@ func (c *DBFT) handleDKG(h *types.Header, state *state.StateDB) error {
 		}
 		state = s
 	}
-	epochDuration, err := getEpochDuration(c.ethAPI, state, h)
+	epochDuration, err := getEpochDuration(c.backend, state, h)
 	if err != nil {
 		return fmt.Errorf("failed to fetch epoch duration: %v", err)
 	}
-	sharePeriodDuration, err := getSharePeriodDuration(c.ethAPI, state, h)
+	sharePeriodDuration, err := getSharePeriodDuration(c.backend, state, h)
 	if err != nil {
 		return fmt.Errorf("failed to fetch share period duration: %v", err)
 	}
@@ -106,11 +106,11 @@ func (c *DBFT) handleDKG(h *types.Header, state *state.StateDB) error {
 		}
 		// if indexOfSharing is 0, then selfPvss should be nil
 		// Ref https://github.com/bane-labs/go-ethereum/blob/4c9105ea2bc246729db0540fce2df02074e21087/contracts/solidity/KeyManagement.sol#L113
-		selfPvss, err := getSharePVSS(c.ethAPI, c.dkgSnapshot.Round, uint64(indexOfSharing), state, h)
+		selfPvss, err := getSharePVSS(c.backend, c.dkgSnapshot.Round, uint64(indexOfSharing), state, h)
 		if err != nil {
 			return fmt.Errorf("failed to fetch spvss, err: %v", err)
 		}
-		aggregatedCommitment, err := getAggregatedCommitment(c.ethAPI, c.dkgSnapshot.Round, state, h)
+		aggregatedCommitment, err := getAggregatedCommitment(c.backend, c.dkgSnapshot.Round, state, h)
 		if err != nil {
 			return fmt.Errorf("failed to fetch aggregated commitment, err: %v", err)
 		}
@@ -123,7 +123,7 @@ func (c *DBFT) handleDKG(h *types.Header, state *state.StateDB) error {
 	}
 
 	// If there is not a snapshot of current epoch, then new
-	epochStartHeight, err := getCurrentEpochStartHeight(c.ethAPI, state, h)
+	epochStartHeight, err := getCurrentEpochStartHeight(c.backend, state, h)
 	if err != nil {
 		return fmt.Errorf("failed to fetch current epoch start height: %v", err)
 	}
@@ -192,11 +192,11 @@ func (c *DBFT) handleDKG(h *types.Header, state *state.StateDB) error {
 		}
 		// if indexOfSharing is 0, then selfPvss should be nil
 		// Ref https://github.com/bane-labs/go-ethereum/blob/4c9105ea2bc246729db0540fce2df02074e21087/contracts/solidity/KeyManagement.sol#L113
-		selfPvss, err := getSharePVSS(c.ethAPI, c.dkgSnapshot.Round-1, uint64(indexOfSharing), state, h)
+		selfPvss, err := getSharePVSS(c.backend, c.dkgSnapshot.Round-1, uint64(indexOfSharing), state, h)
 		if err != nil {
 			return fmt.Errorf("failed to fetch spvsses, err: %v", err)
 		}
-		aggregatedCommitment, err := getAggregatedCommitment(c.ethAPI, c.dkgSnapshot.Round-1, state, h)
+		aggregatedCommitment, err := getAggregatedCommitment(c.backend, c.dkgSnapshot.Round-1, state, h)
 		if err != nil {
 			return fmt.Errorf("failed to fetch aggregated commitment, err: %v", err)
 		}
@@ -209,7 +209,7 @@ func (c *DBFT) handleDKG(h *types.Header, state *state.StateDB) error {
 	// DKG checkpoint handling, also syncs the dkg process if keystore is out-of-date
 	if currentHeight >= shareStartHeight && !c.dkgSnapshot.ShareTasked {
 		// Send share and reshare tx when currentHeight == shareStartHeight
-		c.dkgSnapshot.PendingCNs, err = getPendingConsensus(c.ethAPI, state, h)
+		c.dkgSnapshot.PendingCNs, err = getPendingConsensus(c.backend, state, h)
 		if err != nil {
 			return fmt.Errorf("failed to fetch pending consensus: %v", err)
 		}
@@ -223,7 +223,7 @@ func (c *DBFT) handleDKG(h *types.Header, state *state.StateDB) error {
 		if currentHeight < recoverStartHeight {
 			// If is a member of pending consensus
 			indexOfSharing := slices.Index(c.dkgSnapshot.PendingCNs, amevAddress) + 1
-			receiverMessageKeys, err := getMessagePubkeys(c.ethAPI, c.dkgSnapshot.PendingCNs, state, h)
+			receiverMessageKeys, err := getMessagePubkeys(c.backend, c.dkgSnapshot.PendingCNs, state, h)
 			if err != nil {
 				return fmt.Errorf("failed to get message keys, err: %v", err)
 			}
@@ -246,7 +246,7 @@ func (c *DBFT) handleDKG(h *types.Header, state *state.StateDB) error {
 
 	if currentHeight >= recoverStartHeight && !c.dkgSnapshot.RecoverTasked {
 		// Check isShareReady at height recoverStartHeight
-		ready, err := isShareReady(c.ethAPI, state, h)
+		ready, err := isShareReady(c.backend, state, h)
 		if err != nil {
 			return fmt.Errorf("failed to check if sharing is ready: %v", err)
 		}
@@ -262,7 +262,7 @@ func (c *DBFT) handleDKG(h *types.Header, state *state.StateDB) error {
 				return fmt.Errorf("failed to sync sharing DKG, err: %v", err)
 			}
 		}
-		c.dkgSnapshot.IndexNeedRecover, err = getIndexCurrentNeedRecovering(c.ethAPI, state, h)
+		c.dkgSnapshot.IndexNeedRecover, err = getIndexCurrentNeedRecovering(c.backend, state, h)
 		if err != nil {
 			return fmt.Errorf("failed to fecth index need recovering, err: %v", err)
 		}
@@ -286,7 +286,7 @@ func (c *DBFT) handleDKG(h *types.Header, state *state.StateDB) error {
 				if indexOfResharing > 0 {
 					pubs := make([]*ecies.PublicKey, len(c.dkgSnapshot.IndexNeedRecover))
 					for i, index := range c.dkgSnapshot.IndexNeedRecover {
-						pubs[i], err = getMessagePubkey(c.ethAPI, c.dkgSnapshot.PendingCNs[index-1], state, h)
+						pubs[i], err = getMessagePubkey(c.backend, c.dkgSnapshot.PendingCNs[index-1], state, h)
 						if err != nil {
 							return fmt.Errorf("failed to get message keys for recovering, err: %v", err)
 						}
@@ -309,7 +309,7 @@ func (c *DBFT) handleDKG(h *types.Header, state *state.StateDB) error {
 			if err := c.syncRecoveredSecrets(c.dkgSnapshot, indexOfSharing, state, h); err != nil {
 				return fmt.Errorf("failed to sync recovering DKG, err: %v", err)
 			}
-			receiverMessageKeys, err := getMessagePubkeys(c.ethAPI, c.dkgSnapshot.PendingCNs, state, h)
+			receiverMessageKeys, err := getMessagePubkeys(c.backend, c.dkgSnapshot.PendingCNs, state, h)
 			if err != nil {
 				return fmt.Errorf("failed to fecth message keys, err: %v", err)
 			}
@@ -412,11 +412,11 @@ func (c *DBFT) syncThisRoundSecrets(snap *Snapshot, selfIndex int, state *state.
 
 func (c *DBFT) downloadAndReceiveShare(round uint64, fromIndex int, selfIndex int, state *state.StateDB, header *types.Header) error {
 	// Call ReceiveSecretShare
-	shareMsgs, err := getShareMsgs(c.ethAPI, round, uint64(fromIndex), state, header)
+	shareMsgs, err := getShareMsgs(c.backend, round, uint64(fromIndex), state, header)
 	if err != nil {
 		return err
 	}
-	spvss, err := getSharePVSS(c.ethAPI, round, uint64(fromIndex), state, header)
+	spvss, err := getSharePVSS(c.backend, round, uint64(fromIndex), state, header)
 	if err != nil {
 		return err
 	}
@@ -429,13 +429,13 @@ func (c *DBFT) downloadAndReceiveShare(round uint64, fromIndex int, selfIndex in
 
 func (c *DBFT) downloadAndReceiveReshare(round uint64, fromIndex int, selfIndex int, state *state.StateDB, header *types.Header) error {
 	// Call ReceiveSecretReshare
-	rpvss, err := getResharePVSS(c.ethAPI, round, uint64(fromIndex), state, header)
+	rpvss, err := getResharePVSS(c.backend, round, uint64(fromIndex), state, header)
 	if err != nil {
 		return err
 	}
 	// Only receive reshare has value
 	if len(rpvss) > 0 {
-		reshareMsgs, err := getReshareMsgs(c.ethAPI, round, uint64(fromIndex), state, header)
+		reshareMsgs, err := getReshareMsgs(c.backend, round, uint64(fromIndex), state, header)
 		if err != nil {
 			return err
 		}
@@ -449,12 +449,12 @@ func (c *DBFT) downloadAndReceiveReshare(round uint64, fromIndex int, selfIndex 
 
 // syncRecoveredSecrets downloads DKG recoverings and related PVSS, and sends to keystore
 func (c *DBFT) syncRecoveredSecrets(snap *Snapshot, selfIndex int, state *state.StateDB, header *types.Header) error {
-	pvss, err := getSharePVSS(c.ethAPI, snap.Round-1, uint64(selfIndex), state, header)
+	pvss, err := getSharePVSS(c.backend, snap.Round-1, uint64(selfIndex), state, header)
 	if err != nil {
 		return err
 	}
 	for i := range snap.CurrentCNs {
-		msg, err := getRecoverMsgs(c.ethAPI, snap.Round, uint64(i+1), uint64(selfIndex-1), state, header)
+		msg, err := getRecoverMsgs(c.backend, snap.Round, uint64(i+1), uint64(selfIndex-1), state, header)
 		if err != nil {
 			return err
 		}
@@ -472,12 +472,12 @@ func (c *DBFT) syncRecoveredSecrets(snap *Snapshot, selfIndex int, state *state.
 func (c *DBFT) syncRecoveredReshares(snap *Snapshot, selfIndex int, state *state.StateDB, header *types.Header) error {
 	for _, index := range snap.IndexNeedRecover {
 		// Call ReceiveSecretReshare
-		rpvss, err := getResharePVSS(c.ethAPI, snap.Round, index, state, header)
+		rpvss, err := getResharePVSS(c.backend, snap.Round, index, state, header)
 		if err != nil {
 			return err
 		}
 		if len(rpvss) > 0 {
-			reshareMsgs, err := getReshareMsgs(c.ethAPI, snap.Round, index, state, header)
+			reshareMsgs, err := getReshareMsgs(c.backend, snap.Round, index, state, header)
 			if err != nil {
 				return err
 			}
@@ -576,55 +576,60 @@ func (c *DBFT) taskReshareRecover(receiverMessageKeys []*ecies.PublicKey, start 
 }
 
 // getCurrentConsensus returns an address list of current CNs
-func getCurrentConsensus(api *ethapi.BlockChainAPI, state *state.StateDB, header *types.Header) ([]common.Address, error) {
-	result, err := readFromContract(api, systemcontracts.GovernanceProxyHash, systemcontracts.GovernanceABI, state, header, "getCurrentConsensus")
+func getCurrentConsensus(backend *ethapi.Backend, state *state.StateDB, header *types.Header) ([]common.Address, error) {
+	var result []common.Address
+	err := readFromContract(&result, backend, systemcontracts.GovernanceProxyHash, systemcontracts.GovernanceABI, state, header, "getCurrentConsensus")
 	if err != nil {
 		return nil, err
 	}
-	return result.([]common.Address), nil
+	return result, nil
 }
 
 // getPendingConsensus returns an address list of pending CNs
-func getPendingConsensus(api *ethapi.BlockChainAPI, state *state.StateDB, header *types.Header) ([]common.Address, error) {
-	result, err := readFromContract(api, systemcontracts.GovernanceProxyHash, systemcontracts.GovernanceABI, state, header, "getPendingConsensus")
+func getPendingConsensus(backend *ethapi.Backend, state *state.StateDB, header *types.Header) ([]common.Address, error) {
+	var result []common.Address
+	err := readFromContract(&result, backend, systemcontracts.GovernanceProxyHash, systemcontracts.GovernanceABI, state, header, "getPendingConsensus")
 	if err != nil {
 		return nil, err
 	}
-	return result.([]common.Address), nil
+	return result, nil
 }
 
 // getSharePeriodDuration returns a number of blocks as the duration of each sharing period
-func getSharePeriodDuration(api *ethapi.BlockChainAPI, state *state.StateDB, header *types.Header) (uint64, error) {
-	result, err := readFromContract(api, systemcontracts.GovernanceProxyHash, systemcontracts.GovernanceABI, state, header, "sharePeriodDuration")
+func getSharePeriodDuration(backend *ethapi.Backend, state *state.StateDB, header *types.Header) (uint64, error) {
+	var result *big.Int
+	err := readFromContract(&result, backend, systemcontracts.GovernanceProxyHash, systemcontracts.GovernanceABI, state, header, "sharePeriodDuration")
 	if err != nil {
 		return 0, err
 	}
-	return result.(*big.Int).Uint64(), nil
+	return result.Uint64(), nil
 }
 
 // getEpochDuration returns a number of blocks as the duration of each governanace epoch
-func getEpochDuration(api *ethapi.BlockChainAPI, state *state.StateDB, header *types.Header) (uint64, error) {
-	result, err := readFromContract(api, systemcontracts.GovernanceProxyHash, systemcontracts.GovernanceABI, state, header, "epochDuration")
+func getEpochDuration(backend *ethapi.Backend, state *state.StateDB, header *types.Header) (uint64, error) {
+	var result *big.Int
+	err := readFromContract(&result, backend, systemcontracts.GovernanceProxyHash, systemcontracts.GovernanceABI, state, header, "epochDuration")
 	if err != nil {
 		return 0, err
 	}
-	return result.(*big.Int).Uint64(), nil
+	return result.Uint64(), nil
 }
 
 // getCurrentEpochStartHeight returns the block height when the current governanace epoch starts
-func getCurrentEpochStartHeight(api *ethapi.BlockChainAPI, state *state.StateDB, header *types.Header) (uint64, error) {
-	result, err := readFromContract(api, systemcontracts.GovernanceProxyHash, systemcontracts.GovernanceABI, state, header, "currentEpochStartHeight")
+func getCurrentEpochStartHeight(backend *ethapi.Backend, state *state.StateDB, header *types.Header) (uint64, error) {
+	var result *big.Int
+	err := readFromContract(&result, backend, systemcontracts.GovernanceProxyHash, systemcontracts.GovernanceABI, state, header, "currentEpochStartHeight")
 	if err != nil {
 		return 0, err
 	}
-	return result.(*big.Int).Uint64(), nil
+	return result.Uint64(), nil
 }
 
 // getMessagePubkeys returns the message keys of input address list
-func getMessagePubkeys(api *ethapi.BlockChainAPI, addrs []common.Address, state *state.StateDB, header *types.Header) ([]*ecies.PublicKey, error) {
+func getMessagePubkeys(backend *ethapi.Backend, addrs []common.Address, state *state.StateDB, header *types.Header) ([]*ecies.PublicKey, error) {
 	result := make([]*ecies.PublicKey, len(addrs))
 	for i, addr := range addrs {
-		pub, err := getMessagePubkey(api, addr, state, header)
+		pub, err := getMessagePubkey(backend, addr, state, header)
 		if err != nil {
 			return nil, err
 		}
@@ -634,16 +639,17 @@ func getMessagePubkeys(api *ethapi.BlockChainAPI, addrs []common.Address, state 
 }
 
 // getMessagePubkey returns the message key of input address
-func getMessagePubkey(api *ethapi.BlockChainAPI, addr common.Address, state *state.StateDB, header *types.Header) (*ecies.PublicKey, error) {
-	result, err := readFromContract(api, systemcontracts.KeyManagementProxyHash, systemcontracts.KeyManagementABI, state, header, "messagePubkeys", addr)
+func getMessagePubkey(backend *ethapi.Backend, addr common.Address, state *state.StateDB, header *types.Header) (*ecies.PublicKey, error) {
+	var result string
+	err := readFromContract(&result, backend, systemcontracts.KeyManagementProxyHash, systemcontracts.KeyManagementABI, state, header, "messagePubkeys", addr)
 	if err != nil {
 		return nil, err
 	}
-	if len(result.(string)) == 0 {
+	if len(result) == 0 {
 		err = errors.New("messagePubkey is empty, addr: " + addr.String())
 		return nil, err
 	}
-	key, err := crypto.UnmarshalPubkey([]byte(result.(string)))
+	key, err := crypto.UnmarshalPubkey([]byte(result))
 	if err != nil {
 		return nil, err
 	}
@@ -651,97 +657,106 @@ func getMessagePubkey(api *ethapi.BlockChainAPI, addr common.Address, state *sta
 }
 
 // getIndexCurrentNeedRecovering returns an array of DKG index that needs recover
-func getIndexCurrentNeedRecovering(api *ethapi.BlockChainAPI, state *state.StateDB, header *types.Header) ([]uint64, error) {
-	result, err := readFromContract(api, systemcontracts.KeyManagementProxyHash, systemcontracts.KeyManagementABI, state, header, "indexCurrentNeedRecovering")
+func getIndexCurrentNeedRecovering(backend *ethapi.Backend, state *state.StateDB, header *types.Header) ([]uint64, error) {
+	var result []*big.Int
+	err := readFromContract(&result, backend, systemcontracts.KeyManagementProxyHash, systemcontracts.KeyManagementABI, state, header, "indexCurrentNeedRecovering")
 	if err != nil {
 		return nil, err
 	}
 	var indexs []uint64
-	for _, item := range result.([]*big.Int) {
+	for _, item := range result {
 		indexs = append(indexs, item.Uint64())
 	}
 	return indexs, nil
 }
 
 // isShareReady checks if the DKG sharing is 100% uploaded
-func isShareReady(api *ethapi.BlockChainAPI, state *state.StateDB, header *types.Header) (bool, error) {
-	result, err := readFromContract(api, systemcontracts.KeyManagementProxyHash, systemcontracts.KeyManagementABI, state, header, "isShareReady")
+func isShareReady(backend *ethapi.Backend, state *state.StateDB, header *types.Header) (bool, error) {
+	var result bool
+	err := readFromContract(&result, backend, systemcontracts.KeyManagementProxyHash, systemcontracts.KeyManagementABI, state, header, "isShareReady")
 	if err != nil {
 		return false, err
 	}
-	return result.(bool), nil
+	return result, nil
 }
 
-func getReshareMsgs(api *ethapi.BlockChainAPI, round, index uint64, state *state.StateDB, header *types.Header) ([][]byte, error) {
-	result, err := readFromContract(api, systemcontracts.KeyManagementProxyHash, systemcontracts.KeyManagementABI,
+func getReshareMsgs(backend *ethapi.Backend, round, index uint64, state *state.StateDB, header *types.Header) ([][]byte, error) {
+	var result [][]byte
+	err := readFromContract(&result, backend, systemcontracts.KeyManagementProxyHash, systemcontracts.KeyManagementABI,
 		state, header, "getReshareMsgs", big.NewInt(int64(round)), big.NewInt(int64(index)))
 	if err != nil {
 		return nil, err
 	}
-	return result.([][]byte), nil
+	return result, nil
 }
 
-func getResharePVSS(api *ethapi.BlockChainAPI, round, index uint64, state *state.StateDB, header *types.Header) ([]byte, error) {
-	result, err := readFromContract(api, systemcontracts.KeyManagementProxyHash, systemcontracts.KeyManagementABI,
+func getResharePVSS(backend *ethapi.Backend, round, index uint64, state *state.StateDB, header *types.Header) ([]byte, error) {
+	var result []byte
+	err := readFromContract(&result, backend, systemcontracts.KeyManagementProxyHash, systemcontracts.KeyManagementABI,
 		state, header, "rpvsses", big.NewInt(int64(round)), big.NewInt(int64(index)))
 	if err != nil {
 		return nil, err
 	}
-	return result.([]byte), nil
+	return result, nil
 }
 
-func getShareMsgs(api *ethapi.BlockChainAPI, round, index uint64, state *state.StateDB, header *types.Header) ([][]byte, error) {
-	result, err := readFromContract(api, systemcontracts.KeyManagementProxyHash, systemcontracts.KeyManagementABI,
+func getShareMsgs(backend *ethapi.Backend, round, index uint64, state *state.StateDB, header *types.Header) ([][]byte, error) {
+	var result [][]byte
+	err := readFromContract(&result, backend, systemcontracts.KeyManagementProxyHash, systemcontracts.KeyManagementABI,
 		state, header, "getShareMsgs", big.NewInt(int64(round)), big.NewInt(int64(index)))
 	if err != nil {
 		return nil, err
 	}
-	return result.([][]byte), nil
+	return result, nil
 }
 
-func getSharePVSS(api *ethapi.BlockChainAPI, round, index uint64, state *state.StateDB, header *types.Header) ([]byte, error) {
-	result, err := readFromContract(api, systemcontracts.KeyManagementProxyHash, systemcontracts.KeyManagementABI,
+func getSharePVSS(backend *ethapi.Backend, round, index uint64, state *state.StateDB, header *types.Header) ([]byte, error) {
+	var result []byte
+	err := readFromContract(&result, backend, systemcontracts.KeyManagementProxyHash, systemcontracts.KeyManagementABI,
 		state, header, "spvsses", big.NewInt(int64(round)), big.NewInt(int64(index)))
 	if err != nil {
 		return nil, err
 	}
-	return result.([]byte), nil
+	return result, nil
 }
 
-func getRecoverMsgs(api *ethapi.BlockChainAPI, round, senderIndex, arrIndex uint64, state *state.StateDB, header *types.Header) ([]byte, error) {
-	result, err := readFromContract(api, systemcontracts.KeyManagementProxyHash, systemcontracts.KeyManagementABI,
+func getRecoverMsgs(backend *ethapi.Backend, round, senderIndex, arrIndex uint64, state *state.StateDB, header *types.Header) ([]byte, error) {
+	var result []byte
+	err := readFromContract(&result, backend, systemcontracts.KeyManagementProxyHash, systemcontracts.KeyManagementABI,
 		state, header, "recoverMsgs", big.NewInt(int64(round)), big.NewInt(int64(senderIndex)), big.NewInt(int64(arrIndex)))
 	if err != nil {
 		return nil, err
 	}
-	return result.([]byte), nil
+	return result, nil
 }
 
-func getAggregatedCommitment(api *ethapi.BlockChainAPI, round uint64, state *state.StateDB, header *types.Header) ([]byte, error) {
-	result, err := readFromContract(api, systemcontracts.KeyManagementProxyHash, systemcontracts.KeyManagementABI,
+func getAggregatedCommitment(backend *ethapi.Backend, round uint64, state *state.StateDB, header *types.Header) ([]byte, error) {
+	var result []byte
+	err := readFromContract(&result, backend, systemcontracts.KeyManagementProxyHash, systemcontracts.KeyManagementABI,
 		state, header, "aggregatedCommitments", big.NewInt(int64(round)))
 	if err != nil {
 		return nil, err
 	}
-	return result.([]byte), nil
+	return result, nil
 }
 
-func getRoundNumber(api *ethapi.BlockChainAPI, state *state.StateDB, header *types.Header) (uint64, error) {
-	result, err := readFromContract(api, systemcontracts.KeyManagementProxyHash, systemcontracts.KeyManagementABI,
+func getRoundNumber(backend *ethapi.Backend, state *state.StateDB, header *types.Header) (uint64, error) {
+	var result *big.Int
+	err := readFromContract(&result, backend, systemcontracts.KeyManagementProxyHash, systemcontracts.KeyManagementABI,
 		state, header, "roundNumber")
 	if err != nil {
 		return 0, err
 	}
-	return result.(*big.Int).Uint64(), nil
+	return result.Uint64(), nil
 }
 
-func readFromContract(api *ethapi.BlockChainAPI, contract common.Address, contractAbi abi.ABI, state *state.StateDB, header *types.Header, method string, args ...interface{}) (interface{}, error) {
-	if api == nil {
-		return nil, errors.New("eth blockchain API is not initialized, DKG can't function properly")
+func readFromContract(res interface{}, backend *ethapi.Backend, contract common.Address, contractAbi abi.ABI, state *state.StateDB, header *types.Header, method string, args ...interface{}) error {
+	if backend == nil {
+		return errors.New("eth API backend is not initialized, DKG can't function properly")
 	}
 	data, err := contractAbi.Pack(method, args...)
 	if err != nil {
-		return nil, fmt.Errorf("failed to pack '%s': %v", method, err)
+		return fmt.Errorf("failed to pack '%s': %v", method, err)
 	}
 	msgData := hexutil.Bytes(data)
 	gas := hexutil.Uint64(50_000_000) // More than enough for validators call processing.
@@ -754,15 +769,11 @@ func readFromContract(api *ethapi.BlockChainAPI, contract common.Address, contra
 	ctx, cancel := context.WithCancel(context.Background())
 	// Cancel when we are finished consuming integers.
 	defer cancel()
-	result, err := api.CallAtState(ctx, txArgs, state, header)
+	result, err := ethapi.DoCallAtState(ctx, *backend, txArgs, state, header, nil, nil, 0, 0)
 	if err != nil {
-		return nil, fmt.Errorf("failed to call at state '%s': %v", method, err)
+		return fmt.Errorf("failed to call at state '%s': %v", method, err)
 	}
-	results, err := contractAbi.Unpack(method, result)
-	if err != nil {
-		return nil, fmt.Errorf("failed to unpack result: %v", err)
-	}
-	return results[0], nil
+	return unpackContractExecutionResult(res, result, contractAbi, method)
 }
 
 func sendReshareTx(api *ethapi.TransactionAPI, from common.Address, pvss []byte, messages [][]byte) (*common.Hash, error) {

--- a/eth/backend.go
+++ b/eth/backend.go
@@ -295,10 +295,7 @@ func New(stack *node.Node, config *ethconfig.Config) (*Ethereum, error) {
 	}
 	eth.dbftSrv = dbftproto.New(ethapi.NewBlockChainAPI(eth.APIBackend), onPayload, isExtensibleAllowed)
 	if bft != nil {
-		ethAPI := ethapi.NewBlockChainAPI(eth.APIBackend)
-		bft.WithEthAPI(ethAPI)
-		txAPI := ethapi.NewTransactionAPI(eth.APIBackend, new(ethapi.AddrLocker))
-		bft.WithTxAPI(txAPI)
+		bft.WithEthAPIBackend(eth.APIBackend)
 		bft.WithBroadcast(eth.dbftSrv.BroadcastMessage)
 		bft.WithTxPool(eth.TxPool())
 		bft.WithRequestTxs(eth.handler.BroadcastRequestTxs)

